### PR TITLE
Automated GitHub Release

### LINKED ISSUE #43393: Add Makefile target for creating a Github release
Our release process is largely automated with the following Makefile commands driving it:

- [make update-version](https://github.com/gravitational/teleport/blob/5e25f0192e0d77148dfb66eb8bd4025ecf11f209/Makefile#L1209): updates release version in various scripts/files
- [make update-tag](https://github.com/gravitational/teleport/blob/5e25f0192e0d77148dfb66eb8bd4025ecf11f209/Makefile#L1237): tags the commit and pushes tags to Github
- [make tag-build](https://github.com/gravitational/teleport/blob/5e25f0192e0d77148dfb66eb8bd4025ecf11f209/Makefile#L1271): kicks off GHA workflow for building release assets
- [make tag-publish](https://github.com/gravitational/teleport/blob/5e25f0192e0d77148dfb66eb8bd4025ecf11f209/Makefile#L1291): kicks off GHA workflow for promoting built assets to production repositories

The part with creating a [Github release](https://github.com/gravitational/teleport/releases) is still manual though.

Let's create another Makefile target, say `make create-github-release` that uses Github CLI tool to create a Github release. It should populate the release description appropriately, with the release's changelog extracted from `CHANGELOG.md` file and corresponding "Download" and "Teleport Plugins" sections. Maybe have a template with the release description which we can easily modify if needed.

When creating a release, keep in mind that only the release from the latest "release branch" (e.g. `branch/v16` at this moment) should be set as "latest".

### PR DESCRIPTION
**Pull request by doggydogworld**
Closes #43393 

# Purpose

Automate the release process by introducing a new Makefile target `create-github-release` that is populated from `CHANGELOG.md`.

# Implementation

## release-notes

Introduced a new `release-notes` generation tool that parses through the `CHANGELOG.md` to populate the description of the release. Intended to be used with the `gh` CLI to create releases.

```shell
release-notes $VERSION CHANGELOG.md | gh release create \
    -t "Teleport $VERSION" \
    --latest=false \
    --target=$BRANCH \
    --verify-tag \
    -F - \
```

## Makefile target

Introduces a new Makefile target `create-github-release` that calls the new `release-notes` tooling.

This is by default configured to not set the release as latest. This could be a manual step after the release by just clicking on the checkbox or a specific configuration set in the latest release branch.

# Example Usage

```console
$ make create-release-notes
https://github.com/gravitational/teleport/releases/tag/v16.0.5
```

### CODE REVIEW

File: build.assets/tooling/cmd/release-notes/release_notes.go
Comment (camscale) [inline build.assets/tooling/cmd/release-notes/release_notes.go:58]: This should take an `io.Reader` as a parameter, not `*os.File` as `io.Reader` is all that it needs. It then allows testing with other simulated input not from files. Same for `generateReleaseNotes()` above.

File: build.assets/tooling/cmd/release-notes/release_notes_test.go
Comment (camscale) [inline build.assets/tooling/cmd/release-notes/release_notes_test.go:11]: This could do with some additional tests to test the failure conditions as well.

**Review by camscale — COMMENTED** [review]
A bunch of my comments are about abstraction style, and I'm happy for you to take them as advisory only. To explain a little deeper, one of the reasons I like Go is that it emphasises a simple style, and I've really come to appreciate the simplicity of idiomatic Go that I can read Go code faster than other languages, effectively scanning it and it making sense. Part of that is not using unnecessary abstractions, as an abstraction has a cost when understanding code.

A function is an abstraction, so I really try to avoid single line functions called only once or twice, and instead just open-code what the function is doing. I often find that makes it clearer (although sometimes it's a line call).

Another bit that can be simpler us the use of regular expressions. We only need simple string matching here, so IMHO regexes are overkill in this case.

When I was reading your `parseMD` method, I was jumping up and down the file to see what the functions mean - for example `isEmpty()` - is not really just checking empty, it is checking for empty or whitespace only. The regexes had me bouncing around the file only to learn it was a simple string comparison being done.

This is how I would have written the method:
```
func (r *releaseNotesGenerator) parseMD(md io.Reader) (string, error) {
        sc := bufio.NewScanner(md)

        // Skip until first double header which should be version
        heading := ""
        for sc.Scan() {
                if strings.HasPrefix(sc.Text(), "## ") {
                        heading = sc.Text()
                        break
                }
        }
        if sc.Err() != nil {
                return "", sc.Err()
        }
        if heading == "" {
                return "", trace.BadParameter("no second-level heading found in changelog")
        }

        parts := strings.SplitN(heading, " ", 3)
        // There will always be a parts[1] because there must be at least one space in the heading
        if parts[1] != r.releaseVersion {
                return "", trace.BadParameter("changelog version number did not match expected version number: %s != %s", parts[1], r.releaseVersion)
        }

        // Write everything until next header to buffer
        var buff bytes.Buffer
        for sc.Scan() && !strings.HasPrefix(sc.Text(), "## ") {
                if _, err := fmt.Fprintln(&buff, sc.Text()); err != nil {
                        return "", trace.Wrap(err)
                }
        }
        if sc.Err() != nil {
                return "", sc.Err()
        }

        return strings.TrimSpace(buff.String()), nil
}
```

The method itself is a little longer, but as I scan my eyes down that, it is clear what's going on without needing to reference other parts of the file. I'm also mindful of duplicating `"## "`, but it's only twice and the directness of it I think offsets the duplication and additional burden should we decide to make it a level-three heading in future for example.

Anyway, there's a couple of errors not being checked that should be fixed, and I think also some tests for the error cases too. Happy to debate the abstraction style points or for you to just push ahead as-is if you feel strongly about it.

**Comment by camscale** [comment]
I just noticed I didn't use `trace` on `sc.Err()`. Perhaps do that.

File: build.assets/tooling/cmd/release-notes/release_notes.go
Comment (doggydogworld) [reply build.assets/tooling/cmd/release-notes/release_notes.go:69]: I did the regex as a way to just ignore any potential trailing spaces in the header.

e.g. with extra space which is valid Markdown and will render without the extra spaces.
```
##  16.0.4
```
This would error with a simple split. I don't think this is likely to happen since the changelog is auto generated but I wouldn't expect to fail because of that.

It's originally why I did the double header regex to account for the trailing spaces but then I thought maybe it would be better to just match the version. Regex for semver matching is heinous though so I do think it would be better to do string matching.

File: build.assets/tooling/cmd/release-notes/release_notes.go
Comment (doggydogworld) [reply build.assets/tooling/cmd/release-notes/release_notes.go]: Yeah just had a brain fart when trying to remember this.

File: build.assets/tooling/cmd/release-notes/release_notes.go
Comment (doggydogworld) [reply build.assets/tooling/cmd/release-notes/release_notes.go]: Yeah you're right this was for trailing and leading spaces. Went straight to testing immediately after and forgot to clean this up. This is implemented now.

File: Makefile
Comment (doggydogworld) [reply Makefile]: Added a flag to enable this. Defaults to false.

```
make create-github-release LATEST=true
```

File: Makefile
Comment (fheinecke) [inline Makefile:1708]: Why does this always target master?

File: build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl
Comment (fheinecke) [inline build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl:1]: Somebody from docs team should review this. The downloads page pulls from the GH release notes and parses it, so any changes to the structure/formatting should be reviewed by them.

File: build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl
Comment (fheinecke) [inline build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl:20]: It might be convenient to build this via Go, given a slice/map/struct/whatever of plugin information. This would make it easier to add new plugins assets in the future.

I've had a few discussions with a couple of teams about adding smoke tests for the docs/release integration. This approach may make adding these tests easier in the future.

Comment (fheinecke) [inline build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl:20]: This may be out of scope, but the links here are missing several architectures. We've added arm64 binaries for linux, and I think we now publish both amd64 and arm64 for mac, as well as universal for some (but not all) assets.

**Review by fheinecke — COMMENTED** [review]
IMO this should move to shared-workflows, possibly adding some generalization for it as well. This tool + changelog generator tool would be very helpful in some of our smaller open source projects.

File: build.assets/tooling/cmd/release-notes/main.go
Comment (r0mant) [inline build.assets/tooling/cmd/release-notes/main.go:1]: Can you add license headers to .go files?

File: build.assets/tooling/cmd/release-notes/release_notes.go
Comment (r0mant) [inline build.assets/tooling/cmd/release-notes/release_notes.go:53]: Do we have to implement markdown parsing "by hand"? Is there no library we can use?

It's fine either way as there's one less dependency but if we can do it with a couple of lines by using an existing library it'd be worth it IMO.

File: build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl
Comment (r0mant) [reply build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl:1]: It's fine, it's the same format as we've always used. We can tweak later if needed.

**Review by r0mant — APPROVED** [review]
lgtm if this works as expected. Once this merges, do you mind updating our release process checklist on Notion to replace the manual release creation step with this new Makefile target?

File: Makefile
Comment (doggydogworld) [reply Makefile:1708]: Removed this. I wasn't too sure what this was for but by default it uses the default branch. Figured out it's meant for creating the tag if the tag doesn't already exist from the given tag. Since the expectation is for the tag to already be created it's not necessary. I believe it will still default to `master` but it isn't used.

File: build.assets/tooling/cmd/release-notes/release_notes.go
Comment (camscale) [reply build.assets/tooling/cmd/release-notes/release_notes.go:53]: There is a library that could be used: https://pkg.go.dev/rsc.io/markdown - this is the basis of what is used for parsing Go doc comments now that they have added more markdown syntax. But it seems overkill for the basic parsing we're doing here. I think it will end up larger and more complicated because you still need to iterate the structure to get the bits between the headings, but now there's a bunch of new types to deal with. But it might be worth a try to see what it looks like - I might be wrong.

**Review by camscale — APPROVED** [review]
Just a couple of minor things. LGTM

File: build.assets/tooling/cmd/release-notes/release_notes.go
Comment (doggydogworld) [reply build.assets/tooling/cmd/release-notes/release_notes.go:53]: I did take a look at some libraries but they were all meant for rendering and extensibility of Markdown and not so much about searching. So instead of walking lines in a file it would be walking AST nodes. It might be less lines but I thought it might be more complicated.

I can give it a try with one or two to see if it's easier to use.

File: build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl
Comment (doggydogworld) [reply build.assets/tooling/cmd/release-notes/template/release-notes.md.tmpl:20]: I considered this but I thought it might be out of scope. I wanted to be more opinionated on how the changelog is parsed and to make it consistent with the releases we have out now for the first iteration. I was worried that structure that I would decide for that plugin information would be too distracting.

But you're right it would be more convenient to have say a package that just contains information on artifacts.

**Pull request merged**
